platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/TimeLimitedMatcherFactory.java

```diff
@@ -0,0 +1,148 @@
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.platform.util;
+
+import java.time.Duration;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Factory class used to generate time-limited Matcher instances.
+ *
+ * <p>If the regex runs for too long, this matcher will throw a RuntimeException - specifically a
+ * {@link RegExpTimeoutException} that will throw if in use after their timeout.
+ *
+ * <p>Inspired by <a
+ * href="https://www.exratione.com/2017/06/preventing-unbounded-regular-expression-operations-in-java/">
+ * this post at ExRatione</a>
+ */
+public class TimeLimitedMatcherFactory {
+  // Limit regular expression matching to five seconds. Any longer indicates an expression that is
+  // likely to spin indefinitely.
+  private static final Duration RE_DURATION = Duration.ofSeconds(5L);
+
+  private TimeLimitedMatcherFactory() {
+    // Prevent construction of factory
+  }
+
+  /**
+   * Generate a Matcher instance that will timeout after more than {@link #RE_DURATION} seconds
+   * after its instantiation.
+   *
+   * <p>Use the instance immediately and then discard it.
+   *
+   * @param pattern The Pattern instance.
+   * @param charSequence The CharSequence to operate on.
+   */
+  public static Matcher matcher(Pattern pattern, CharSequence charSequence) {
+    return matcher(pattern, charSequence, RE_DURATION);
+  }
+
+  /**
+   * Generate a Matcher instance that will timeout after more than {@code duration} after its
+   * instantiation.
+   *
+   * <p>Use the instance immediately and then discard it.
+   *
+   * @param pattern The Pattern instance.
+   * @param charSequence The CharSequence to operate on.
+   * @param duration Throw after this timeout is reached.
+   */
+  public static Matcher matcher(Pattern pattern, CharSequence charSequence, Duration duration) {
+
+    // Substitute in our exploding CharSequence implementation.
+    if (!(charSequence instanceof TimeLimitedCharSequence)) {
+
+      charSequence =
+          new TimeLimitedCharSequence(
+              charSequence,
+              System.currentTimeMillis() + duration.toMillis(),
+              pattern,
+              charSequence);
+    }
+
+    return pattern.matcher(charSequence);
+  }
+
+  /**
+   * A CharSequence implementation that throws when charAt() is called after a given timeout.
```

Comment:
❓ `throws` what?

Reply:
Oh yeah. I should probably clarify that!